File: sdk/dotnet/src/ServerSentEventGenerator.fs
namespace StarFederation.Datastar

open StarFederation.Datastar.Utility

/// <summary>
/// Converts requests into SSEs, serializes, and sends to sseHandler handlers
/// </summary>
type ServerSentEventGenerator =

    static member send (sseHandler:ISendServerEvent) sse = sseHandler.SendServerEvent sse

    static member mergeFragments (sseHandler, fragments, ?options:MergeFragmentsOptions) =
        let options = defaultArg options MergeFragmentsOptions.defaults
        { EventType = MergeFragments
          Id = options.EventId
          Retry = options.Retry
          DataLines = [|
            if (options.Selector |> ValueOption.isSome) then $"{Consts.DatastarDatalineSelector} {options.Selector |> ValueOption.get |> Selector.value}"
            if (options.MergeMode <> Consts.DefaultFragmentMergeMode) then $"{Consts.DatastarDatalineMergeMode} {options.MergeMode |> Consts.FragmentMergeMode.toString}"
            if (options.SettleDuration <> Consts.DefaultFragmentsSettleDuration) then $"{Consts.DatastarDatalineSettleDuration} {options.SettleDuration.Milliseconds}"
            if (options.UseViewTransition <> Consts.DefaultFragmentsUseViewTransitions) then $"{Consts.DatastarDatalineUseViewTransition} %A{options.UseViewTransition}"
            yield! (fragments |> String.split String.newLines |> Seq.map (fun fragmentLine -> $"{Consts.DatastarDatalineFragments} %s{fragmentLine}"))
            |] }
        |> ServerSentEventGenerator.send sseHandler

    static member removeFragments (sseHandler, selector, ?options:RemoveFragmentsOptions) =
        let options = defaultArg options RemoveFragmentsOptions.defaults
        { EventType = RemoveFragments
          Id = options.EventId
          Retry = options.Retry
          DataLines = [|
            $"{Consts.DatastarDatalineSelector} {selector |> Selector.value}"
            if (options.SettleDuration <> Consts.DefaultFragmentsSettleDuration) then $"{Consts.DatastarDatalineSettleDuration} {options.SettleDuration.Milliseconds}"
            if (options.UseViewTransition <> Consts.DefaultFragmentsUseViewTransitions) then $"{Consts.DatastarDatalineUseViewTransition} %A{options.UseViewTransition}"
            |] }
        |> ServerSentEventGenerator.send sseHandler

    static member mergeSignals (sseHandler, mergeSignals, ?options:MergeSignalsOptions) =
        let options = defaultArg options MergeSignalsOptions.defaults
        { EventType = MergeSignals
          Id = options.EventId
          Retry = options.Retry
          DataLines = [|
            if (options.OnlyIfMissing <> Consts.DefaultMergeSignalsOnlyIfMissing) then $"{Consts.DatastarDatalineOnlyIfMissing} %A{options.OnlyIfMissing}"
            yield! mergeSignals |> Signals.value |> String.split String.newLines |> Seq.map (fun dataLine -> $"{Consts.DatastarDatalineSignals} %s{dataLine}")
            |] }
       |> ServerSentEventGenerator.send sseHandler

    static member removeSignals (sseHandler, paths, ?options:EventOptions) =
        let options = defaultArg options EventOptions.defaults
        let paths' = paths |> Seq.map SignalPath.value |> String.concat " "
        { EventType = RemoveSignals
          Id = options.EventId
          Retry = options.Retry
          DataLines = [| $"{Consts.DatastarDatalinePaths} {paths'}" |] }
        |> ServerSentEventGenerator.send sseHandler

    static member executeScript (sseHandler, script, ?options:ExecuteScriptOptions) =
        let options = defaultArg options ExecuteScriptOptions.defaults
        { EventType = ExecuteScript
          Id = options.EventId
          Retry = options.Retry
          DataLines = [|
            if (options.AutoRemove <> Consts.DefaultExecuteScriptAutoRemove) then $"{Consts.DatastarDatalineAutoRemove} %A{options.AutoRemove}"
            if (not <| Seq.forall2 (=) options.Attributes [| Consts.DefaultExecuteScriptAttributes |] ) then
                yield! options.Attributes |> Seq.map (fun attr -> $"{Consts.DefaultExecuteScriptAttributes} {attr}")
            yield! script |> String.split String.newLines |> Seq.map (fun scriptLine -> $"{Consts.DatastarDatalineScript} %s{scriptLine}")
            |] }
        |> ServerSentEventGenerator.send sseHandler

File: sdk/dotnet/src/Starfederation.Datastar/ServerSentEventGenerator.cs
namespace Starfederation.Datastar
{
    /// <summary>
    /// Converts requests into SSEs, serializes, and sends to ISendServerEvent handlers
    /// </summary>
    public static class ServerSentEventGenerator
    {
        /// <summary>
        /// Sends a server-sent event.
        /// </summary>
        /// <param name="sseHandler">The handler to send the event to.</param>
        /// <param name="sse">The server-sent event to send.</param>
        /// <returns>A task that represents the asynchronous send operation.</returns>
        public static Task Send(ISendServerEvent sseHandler, ServerSentEvent sse)
        {
            return sseHandler.SendServerEventAsync(sse);
        }

        /// <summary>
        /// Sends a merge fragments event.
        /// </summary>
        /// <param name="sseHandler">The handler to send the event to.</param>
        /// <param name="fragments">The fragments to merge.</param>
        /// <param name="options">The options for the merge fragments event.</param>
        /// <returns>A task that represents the asynchronous send operation.</returns>
        public static Task MergeFragments(ISendServerEvent sseHandler, string fragments, MergeFragmentsOptions? options = null)
        {
            options ??= MergeFragmentsOptions.Defaults;

            var dataLines = new System.Collections.Generic.List<string>();
            
            if (options.Selector.HasValue)
            {
                dataLines.Add($"{Consts.DatastarDatalineSelector} {options.Selector.Value.Value}");
            }

            var mergeMode = Consts.FragmentMergeModeHelper.ToString(options.MergeMode);
            if (mergeMode != Consts.FragmentMergeModeHelper.ToString(Consts.DefaultFragmentMergeMode))
            {
                dataLines.Add($"{Consts.DatastarDatalineMergeMode} {mergeMode}");
            }

            dataLines.Add($"{Consts.DatastarDatalineFragments} {fragments}");

            var sse = new ServerSentEvent
            {
                EventType = EventType.MergeFragments,
                Id = options.Selector.HasValue ? Utils.FromEmptyString(options.Selector.Value.Value) : Optional<string>.None,
                Retry = options.Retry,
                DataLines = dataLines.ToArray()
            };

            return Send(sseHandler, sse);
        }

        /// <summary>
        /// Sends a remove fragments event.
        /// </summary>
        /// <param name="sseHandler">The handler to send the event to.</param>
        /// <param name="selector">The selector for the fragments to remove.</param>
        /// <param name="options">The options for the remove fragments event.</param>
        /// <returns>A task that represents the asynchronous send operation.</returns>
        public static Task RemoveFragments(ISendServerEvent sseHandler, Selector selector, RemoveFragmentsOptions? options = null)
        {
            options ??= RemoveFragmentsOptions.Defaults;

            var dataLines = new System.Collections.Generic.List<string>
            {
                $"{Consts.DatastarDatalineSelector} {selector.Value}"
            };

            if (options.UseViewTransition != Consts.DefaultFragmentsUseViewTransitions)
            {
                dataLines.Add($"{Consts.DatastarDatalineUseViewTransition} {options.UseViewTransition.ToString().ToLower()}");
            }

            var sse = new ServerSentEvent
            {
                EventType = EventType.RemoveFragments,
                Id = options.EventId,
                Retry = options.Retry,
                DataLines = dataLines.ToArray()
            };

            return Send(sseHandler, sse);
        }

        /// <summary>
        /// Sends a merge signals event.
        /// </summary>
        /// <param name="sseHandler">The handler to send the event to.</param>
        /// <param name="mergeSignals">The signals to merge.</param>
        /// <param name="options">The options for the merge signals event.</param>
        /// <returns>A task that represents the asynchronous send operation.</returns>
        public static Task MergeSignals(ISendServerEvent sseHandler, Signals mergeSignals, MergeSignalsOptions? options = null)
        {
            options ??= MergeSignalsOptions.Defaults;

            var dataLines = new System.Collections.Generic.List<string>
            {
                $"{Consts.DatastarDatalineSignals} {mergeSignals.Value}"
            };

            if (options.OnlyIfMissing != Consts.DefaultMergeSignalsOnlyIfMissing)
            {
                dataLines.Add($"{Consts.DatastarDatalineOnlyIfMissing} {options.OnlyIfMissing.ToString().ToLower()}");
            }

            var sse = new ServerSentEvent
            {
                EventType = EventType.MergeSignals,
                Id = options.EventId,
                Retry = options.Retry,
                DataLines = dataLines.ToArray()
            };

            return Send(sseHandler, sse);
        }

        /// <summary>
        /// Sends a remove signals event.
        /// </summary>
        /// <param name="sseHandler">The handler to send the event to.</param>
        /// <param name="paths">The signal paths to remove.</param>
        /// <param name="options">The options for the remove signals event.</param>
        /// <returns>A task that represents the asynchronous send operation.</returns>
        public static Task RemoveSignals(ISendServerEvent sseHandler, System.Collections.Generic.IEnumerable<SignalPath> paths, EventOptions? options = null)
        {
            options ??= EventOptions.Defaults;

            var pathsString = string.Join(" ", System.Linq.Enumerable.Select(paths, p => p.Value));

            var sse = new ServerSentEvent
            {
                EventType = EventType.RemoveSignals,
                Id = options.EventId,
                Retry = options.Retry,
                DataLines = new[] { $"{Consts.DatastarDatalinePaths} {pathsString}" }
            };

            return Send(sseHandler, sse);
        }

        /// <summary>
        /// Sends an execute script event.
        /// </summary>
        /// <param name="sseHandler">The handler to send the event to.</param>
        /// <param name="script">The script to execute.</param>
        /// <param name="options">The options for the execute script event.</param>
        /// <returns>A task that represents the asynchronous send operation.</returns>
        public static Task ExecuteScript(ISendServerEvent sseHandler, string script, ExecuteScriptOptions? options = null)
        {
            options ??= ExecuteScriptOptions.Defaults;

            var dataLines = new System.Collections.Generic.List<string>
            {
                $"{Consts.DatastarDatalineScript} {script}"
            };

            var attributes = string.Join(" ", options.Attributes);
            if (attributes != Consts.DefaultExecuteScriptAttributes)
            {
                dataLines.Add($"{Consts.DatastarDatalineAttributes} {attributes}");
            }

            if (options.AutoRemove != Consts.DefaultExecuteScriptAutoRemove)
            {
                dataLines.Add($"{Consts.DatastarDatalineAutoRemove} {options.AutoRemove.ToString().ToLower()}");
            }

            var sse = new ServerSentEvent
            {
                EventType = EventType.ExecuteScript,
                Id = Utils.FromEmptyString(script),
                Retry = options.Retry,
                DataLines = dataLines.ToArray()
            };

            return Send(sseHandler, sse);
        }
    }
}
